File: c#/builder/builder/Director.cs
using System;
using System.Collections;
using System.Xml;

namespace builder
{
	public class Director
	{
		private FormBuilder builder;
		private Stack parents;

		public Director(FormBuilder builder)
		{
			this.builder = builder;
			this.parents = new Stack();
		}

		public void Build(string fileName) {
			XmlReader reader = XmlReader.Create(fileName);
			while (reader.Read()) {
				if (reader.NodeType == XmlNodeType.Element) {
					StartElement(reader.Name, reader.GetAttribute("label"));
				} else if (reader.NodeType == XmlNodeType.EndElement) {
					EndElement(reader.Name);
				}
			}
			reader.Close();
		}

		public void StartElement(string name, string label) {
			if (name.Equals("form")) {
				parents.Push(builder.StartBuilding());
			} else if (name.Equals("group")) {
				parents.Push(builder.StartGroup(parents.Peek(), label));
			} else if (name.Equals("checkbox")) {
				builder.BuildCheckbox(parents.Peek(), label);
			} else if (name.Equals("text")) {
				builder.BuildText(parents.Peek(), label);
			} else if (name.Equals("radio")) {
				parents.Push(builder.StartRadioGroup(parents.Peek()));
			} else if (name.Equals("button")) {
				builder.BuildRadioButton(parents.Peek(), label);
			}
		}

		public void EndElement(string name) {
			if (name.Equals("form")) {
				builder.EndBuilding();
				parents.Pop();
				if (parents.Count != 0) {
					Console.WriteLine("parents is not empty");
				}
			} else if (name.Equals("group")) {
				parents.Pop();
				builder.EndGroup(parents.Peek());
			} else if (name.Equals("radio")) {
				parents.Pop();
				builder.EndRadioGroup(parents.Peek());
			}
		}
	}
}

File: various/oom-lab1/Director.java
package lab1;

import java.util.Stack;

import javax.swing.JFrame;
import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

public class Director extends DefaultHandler {

	private static final String INPUT_FILE = "src/lab1/spec.xml";

	private FormBuilder builder;
	private Stack<Object> parents;

	private Director(FormBuilder builder) {
		this.builder = builder;
		this.parents = new Stack<Object>();
	}

	public void build(String fileName) {
		SAXParserFactory xmlParserFactory = SAXParserFactory.newInstance();
		try {
			SAXParser xmlParser = xmlParserFactory.newSAXParser();
			xmlParser.parse(fileName, this);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public void startDocument() throws SAXException {
		parents.push(builder.startBuilding());
	}

	@Override
	public void endDocument() throws SAXException {
		builder.endBuilding();
		parents.pop();
		assert parents.empty() : "parents stack should have been empty!";
	}

	@Override
	public void startElement(String uri, String localName, String name,
			Attributes attributes) throws SAXException {
		String label = attributes.getValue("label");
		if (name.equalsIgnoreCase("group")) {
			parents.push(builder.startGroup(parents.peek(), label));
		} else if (name.equalsIgnoreCase("radio")) {
			parents.push(builder.startRadioGroup(parents.peek()));
		} else if (name.equalsIgnoreCase("checkbox")) {
			builder.buildCheckbox(parents.peek(), label);
		} else if (name.equalsIgnoreCase("text")) {
			builder.buildText(parents.peek(), label);
		} else if (name.equalsIgnoreCase("button")) {
			builder.buildRadioButton(parents.peek(), label);
		}
	}

	@Override
	public void endElement(String uri, String localName, String name)
			throws SAXException {
		if (name.equalsIgnoreCase("group")) {
			parents.pop();
			builder.endGroup(parents.peek());
		} else if (name.equalsIgnoreCase("radio")) {
			parents.pop();
			builder.endRadioGroup(parents.peek());
		}
	}

	public static void runXmlBuilder() {
		new Director(new XmlFormBuilder()).build(INPUT_FILE);
	}

	public static void runHtmlBuilder() {
		new Director(new HtmlFormBuilder("out.html")).build(INPUT_FILE);
	}

	public static void runGuiBuilder() {
		GuiBuilder guiBuilder = new GuiBuilder();
		new Director(guiBuilder).build(INPUT_FILE);
		JFrame frame =  new JFrame();
		frame.setTitle("GuiBuilder Result");
		frame.getContentPane().add(guiBuilder.getResult());
		frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		frame.pack();
		frame.setVisible(true);
	}

	public static void main(String[] args) {
		//runXmlBuilder();
		//runHtmlBuilder();
		runGuiBuilder();
	}
}
